Add first tutorial lessons (#484)
crystal-lang/crystal-book: docs/tutorials/01_basics/03_math.md

```

Besides the equality operator, there are also comparison operators. They determine the relationship between two values.
As with equality, comparability is also independent of type.

```{.crystal, .crystal-play}
p! 2 > 1,
   1 >= 1,
   1 < 2,
   1 <= 2
```

The main comparison operator is `<=>`, also called *Spaceship operator* for its appearance. It compares its operands and returns a value that is either zero (both operands are equal),
a negative value (the first operand is bigger), or a positive value (the second operand is bigger).

```{.crystal, .crystal-play}
p! 1 <=> 1,
   2 <=> 1,
   1 <=> 2
```

### Operators

Basic arithmetic operations can be performed with operators. Most operators are *binary* (i.e. two operands), and
written in infix notation (i.e. between the operands). Some operators are *unary* (i.e. one operand), and written in prefix
notation (i.e. before the operand).
The value of the expression is the result of the operation.

```{.crystal, .crystal-play}
p! 1 + 1,  # addition
   1 - 1,  # subtraction
   2 * 3,  # multiplication
   2 ** 4, # exponentiation
   2 / 3,  # division
   2 // 3, # floor division
   3 % 2,  # modulus
   -1,     # negation (unary)

>    !4      # faculty (unary)

Reviewer: Not sure what this is...  `!4` == "not 4" == `false`.  I think this might have been going for factorial, but I don't think there's an operator for that?

```

As you can see, the result of most of these operations between integer operands is also an integer value.
The division operator (`/`) is an exception. It always returns a float value. The floor division operator (`//`) however returns an integer value, but it's obviously reduced to integer precision.
An operation between integer and float operands always returns a float value. Otherwise, the return type is usually the type of the first operand.

!!! info
    A full list of operators is available in [the Operator reference](../../syntax_and_semantics/operators.md#arithmetic-operators).

#### Precedence

!!! todo "precedence"

### Number Methods

Some less common math operations are not operators, but named methods.

```{.crystal, .crystal-play}
p! -5.abs,    # absolute value
   4.3.round, # round to nearest integer
   5.even?,   # odd/even check

>    10.gcd    # greatest common divisor

Reviewer: alignment

```suggestion
   10.gcd     # greatest common divisor
```

```

!!! info
    A full list of number methods is available in [the Number API docs](https://crystal-lang.org/api/latest/Number.html) (also check subtypes).

### Math Methods

Some arithmetic methods are not defined on the number types directly, but in the `Math` namespace.

```{.crystal, .crystal-play}
p! Math.cos(1),      # cosine
   Math.sin(1),      # sine

>    Math.tan(1),      # tangens

Reviewer: Typo

```suggestion
   Math.tan(1),      # tangent
```

p! 1, typeof(1)
p! 1.0, typeof(1.0)
p! 100_000, typeof(100_000)
p! 100_000.0, typeof(100_000.0)
```

Float values print with with a decimal point. Integer values don't.

!!! info
    There are quite a few more numeric types, but most of them are intended only for special use cases such as binary protocols,
    specific numeric algorithms and performance optimization. You probably don't need them for every-day programs.

    See [Integer literal reference](../../syntax_and_semantics/literals/integers.md) and [Float literal reference](../../syntax_and_semantics/literals/floats.md)
    for a full reference on all primitive number types and alternative representations.

## Arithmetic

### Equality and Comparison

Numbers of the same numerical value are considered equal regarding the equality operator `==`, independent of their type.

```{.crystal, .crystal-play}
p! 1 == 1,
   1 == 2,
   1.0 == 1,
   -2000.0 == -2000
```

Besides the equality operator, there are also comparison operators. They determine the relationship between two values.
As with equality, comparability is also independent of type.

```{.crystal, .crystal-play}
p! 2 > 1,
   1 >= 1,
   1 < 2,
   1 <= 2
```

> The main comparison operator is `<=>`, also called *Spaceship operator* for its appearance. It compares its operands and returns a value that is either zero (both operands are equal),

Author: The others are `<`, `<=`, `>=`, `>` and mentioned in the code sample directly before.

> The main comparison operator is `<=>`, also called *Spaceship operator* for its appearance. It compares its operands and returns a value that is either zero (both operands are equal),

Reviewer 2: Of course. I was just questioning why the spaceship operator should stand out from among the rest, other than the fact that it's awesome :smile:

> The main comparison operator is `<=>`, also called *Spaceship operator* for its appearance. It compares its operands and returns a value that is either zero (both operands are equal),

Author: Because it combines all of them? Maybe that should be more explicit.

> The main comparison operator is `<=>`, also called *Spaceship operator* for its appearance. It compares its operands and returns a value that is either zero (both operands are equal),

Reviewer 2: Yes, I like the way you reworded it.

   2 / 3,  # division
   2 // 3, # floor division
   3 % 2,  # modulus
   -1,     # negation (unary)
```

As you can see, the result of most of these operations between integer operands is also an integer value.
The division operator (`/`) is an exception. It always returns a float value. The floor division operator (`//`) however returns an integer value, but it's obviously reduced to integer precision.
An operation between integer and float operands always returns a float value. Otherwise, the return type is usually the type of the first operand.

!!! info
    A full list of operators is available in [the Operator reference](../../syntax_and_semantics/operators.md#arithmetic-operators).

#### Precedence

When several operators are combined, the question arises in which order are they executed.
In math there are several rules, like multiplication and division take precedence over addition and subtraction.
Crystal operators implement these precedence rules.

A tool to structure operations are parentheses. An operator expression in parenthesis always takes precedence over external operators.

```{.crystal, .crystal-play}
p! 4 + 5 * 2,
   (4 + 5) * 2
```

!!! info

>     The entire precedence rules are quite extensive in the [the Operator reference](../../syntax_and_semantics/operators.md#operator-precedence).

Reviewer 3: ```suggestion
    All the precedence rules are detailed in the [the Operator reference](../../syntax_and_semantics/operators.md#operator-precedence).
```
